### PR TITLE
Addonmanager upgrade

### PR DESCRIPTION
**Pull request by olkorhon**
## Status
**READY**

## Migrations
NO

## Description
Improves AddonManager so that it provides a more robust interface for adding, registering and removing addons. Most significant change to interface is that most functions now return promises.  

## Todos
- [x] proper commenting
- [x] manual tests
- [x] define sample addon
- [x] initial unittests

## Impacted Areas in Application
* app/index.js
* app/addons/*
* tests/tests_unittests/addons/*

### CODE REVIEW

File: app/index.js [hunk 1]
```diff
@@ -133,25 +133,27 @@ fs.readdirSync(__dirname + '/routes').forEach(function (file) {
 // Bootsrap addons, like default webGUI
 var AddonManager = require('./addons');
 global.AddonManager = new AddonManager(app, server, io);
```
Comment (jupe) [inline app/index.js]: maybe we doesn't need at all - it's not good approach anyway.. if its not used anywhere else just use local variable or perhaps even singleton instance created inside `/addons/index.js`..

File: app/addons/index.js [hunk 2]
```diff
@@ -1,78 +1,339 @@
...
-      }
+// native modules
+const fs = require('fs');
+const path = require('path');
+
+// 3rd party modules
+const logger = require('winston');
+
+process.on('warning', (warn) => {
+  console.log(warn.stack);
```
Comment (jupe) [inline app/addons/index.js]: why not logger instead ? there is also `name` and `message` which would be also useful..

File: app/addons/index.js [hunk 3]
```diff
@@ -1,78 +1,339 @@
...
+const STATES = { introduce: 0, load: 1, instantiate: 2, register: 3, unregister: 4 };
+const PHASES = { inProgress: 0, done: 1, failed: 2 };
+
+function createErrorMessage(pError, pMetaInfo, pSolution) {
+  let combinedError = pError;
+
+  // Append meta data lines
+  for (let i = 0; i < pMetaInfo.length; i += 1) {
+    if (i !== pMetaInfo.length - 1) {
+      combinedError += `\n ├ ${pMetaInfo[i]}`;
```
Comment (jupe) [inline app/addons/index.js]: I don't see any differences with these two prints, what's the point ?

File: app/addons/index.js [hunk 4]
```diff
@@ -1,78 +1,339 @@
...
+});
+
+const STATES = { introduce: 0, load: 1, instantiate: 2, register: 3, unregister: 4 };
+const PHASES = { inProgress: 0, done: 1, failed: 2 };
+
+function createErrorMessage(pError, pMetaInfo, pSolution) {
+  let combinedError = pError;
+
+  // Append meta data lines
+  for (let i = 0; i < pMetaInfo.length; i += 1) {
```
Comment (jupe) [inline app/addons/index.js]: perhaps could be even more simple like: `_.for(pMetaInfo, (info) => { combinedError+=...}`

File: app/addons/index.js [hunk 5]
```diff
@@ -1,78 +1,339 @@
...
+const logger = require('winston');
+
+process.on('warning', (warn) => {
+  console.log(warn.stack);
+});
+
+const STATES = { introduce: 0, load: 1, instantiate: 2, register: 3, unregister: 4 };
+const PHASES = { inProgress: 0, done: 1, failed: 2 };
+
+function createErrorMessage(pError, pMetaInfo, pSolution) {
```
Comment (jupe) [inline app/addons/index.js]: should this be static method for Addon ? is this used anywhere else ?

File: app/addons/index.js [hunk 6]
```diff
@@ -1,78 +1,339 @@
...
+
+  get Status() {
+    const state = Object.keys(STATES).find(key => STATES[key] === this.status.state);
+    const phase = Object.keys(PHASES).find(key => PHASES[key] === this.status.phase);
+
+    return `${state}-${phase}`;
+  }
+
+  get isRegistered() {
+    return this.status.state === STATES.register && this.status.phase === PHASES.done;
```
Comment (jupe) [inline app/addons/index.js]: use two lines like in line 53 so its consistent

File: app/addons/index.js [hunk 7]
```diff
@@ -1,78 +1,339 @@
...
+  }
+
+  get safeToRemove() {
+    return (this.status.state !== STATES.registered) &&
+           (this.status.phase !== PHASES.inProgress);
+  }
+
+  loadModule() {
+    if (this.status.state !== STATES.introduce || this.status.phase === PHASES.inProgress) {
```
Comment (jupe) [inline app/addons/index.js]: this could be also splitted to two lines..

File: app/addons/index.js [hunk 8]
```diff
@@ -1,78 +1,339 @@
...
+  }
+
+  registerAddons() {
+    logger.info('Registering addons...');
+
+    // Promise to register all addons
+    const registerPromises = this.addons.filter((addon) => {
+      return addon.status.state === STATES.load && addon.status.phase === PHASES.done;
+    }).map((addon) => {
+      addon.createInstance(this.app, this.server, this.io);
```
Comment (jupe) [inline app/addons/index.js]: if one create instance fails is it halting rest of plugin registering... ? looks like it does.. maybe it should still try rest of plugins..

File: app/index.js [hunk 9]
```diff
@@ -133,25 +133,27 @@ fs.readdirSync(__dirname + '/routes').forEach(function (file) {
...
-var onListening = function(){
-  var listenurl = (nconf.get('https')?'https':'http:')+'://'+nconf.get('listen')+':'+nconf.get('port');
-  console.log('OpenTMI started on ' +listenurl+ ' in '+ nconf.get('cfg')+ ' mode');
-};
-
-server.listen(nconf.get('port'), nconf.get('listen'));
-server.on('error', onError);
-server.on('listening', onListening);
+global.AddonManager.loadAddons().then(() => {
+  global.AddonManager.registerAddons().then(() => {
```
Comment (jupe) [inline app/index.js]: this could be simplified one identity less...

File: app/addons/index.js (hunk 2 above)
Comment (olkorhon) [inline app/addons/index.js]: That block accidentally made it to the commit, was intended to catch warnings that promises tend to suppress during debugging.

File: app/addons/index.js (hunk 5 above)
Comment (olkorhon) [inline app/addons/index.js]: Cannot really decide where to put it, would like to use it later in other files as well. Storing at addons/index.js as a static method for now.

File: test/tests_unittests/addons/index.js [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+      return Object.getPrototypeOf(AddonManager).constructor._recursiveLoad(addons)
+      .then(() => {
+        expect(loadCalled).to.equal(addons.length, 'load should be called as many times as there are addons.');
+        expect(instanceCalled).to.equal(addons.length, 'create instance should be called as many times as there are addons.');
+        return Promise.resolve();
+      });
+    });
+  });
+
+  describe('_loadAddons', function () {
```
Comment (lorjala) [inline test/tests_unittests/addons/index.js]: maybe remove underscore here? inconsistent...

File: package.json [hunk 11]
```diff
@@ -56,6 +56,7 @@
     "lodash": "^4.17.4",
     "method-override": "^2.3.6",
     "mime": "^1.3.4",
+    "mock-fs": "4.4.1",
```
Comment (jupe) [inline package.json]: this should be dev dep instead?

File: test/tests_unittests/addons/addon.js [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+    it('Status - valid state', function (done) {
+      addon._status = { state: 1, phase: 2 };
+      expect(addon.Status).to.equal('load-failed');
+      done();
+    });
+  });
+
+  describe('isBusy', function () {
+    it('isBusy - true and false case', function (done) {
+      addon._status = { state: 0, phase: 1 };
```
Comment (jupe) [inline test/tests_unittests/addons/addon.js]: plain numbers are not very describe for this kind of thing, some freezed object would be better..

**Pull request merged**